packages/react/src/components/form/ui-state-machine.ts

```diff
@@ -0,0 +1,182 @@
+import type { SlashID } from "@slashid/slashid";
+import type { AuthenticatingState, Send } from "./flow.types";
+import { isFactorOTP, isFactorPassword } from "../../domain/handles";
+import { LogIn, MFA } from "../../domain/types";
+import { PasswordState as PasswordStatus } from "./authenticating/password";
+
+type OTPStatus = "initial" | "input" | "submitting";
+
+type PasswordStatus =
+  | "initial"
+  | "setPassword"
+  | "verifyPassword"
+  | "recoverPassword"
+  | "submitting";
+
+type TOTPStatus =
+  | "initial"
+  | "registerAuthenticator"
+  | "input"
+  | "submitting"
+  | "saveRecoveryCodes";
+
+export type AuthenticatingUIStatus = OTPStatus | PasswordStatus | TOTPStatus;
+
+export interface IUIStateMachine {
+  state: State<AuthenticatingUIStatus>;
+  start(): void;
+}
+
+export interface State<T> {
+  status: T;
+  entry?(): void;
+  exit?(): void;
+}
+
+export interface InputState extends State<AuthenticatingUIStatus> {
+  status: "input";
+  submit: (...values: unknown[]) => void;
+}
+
+type UIStateMachineOpts = {
+  send: Send;
+  sid: SlashID;
+  context: AuthenticatingState["context"];
+  logInFn: LogIn | MFA;
+};
+
+abstract class UIStateMachine<T extends AuthenticatingUIStatus>
+  implements IUIStateMachine
+{
+  state: State<T>;
+  send: Send;
+  sid: SlashID;
+  context: AuthenticatingState["context"];
+  logInFn: LogIn | MFA;
+
+  constructor({ send, sid, context, logInFn }: UIStateMachineOpts) {
+    this.send = send;
+    this.sid = sid;
+    this.state = { status: "NULL_STATUS" as T };
+    this.context = context;
+    this.logInFn = logInFn;
+  }
+
+  start() {
+    if (this.state.status === "initial") {
+      this.state.entry!();
+    }
+  }
+
+  protected transition(state: State<T>) {
+    if (this.state.exit) {
+      this.state.exit();
+    }
+    this.state = state;
+    if (this.state.entry) {
+      this.state.entry();
+    }
+
+    this.send({ type: "sid_login.state_changed", state });
+  }
+}
+
+class OTPUIStateMachine extends UIStateMachine<OTPStatus> {
+  constructor(opts: UIStateMachineOpts) {
+    super(opts);
+    this.state = this.prepareNextState("initial");
+  }
+
+  private prepareNextState(status: OTPStatus): State<OTPStatus> {
```

Comment:
This naming breaks my brain 😅

What does this method do? Seems a bit confusing since it seems to call itself as part of `this.transition` calls.

Usually the API designed around state machines would have a `transition` function that receives an event that is supposed to transition the machine to the next state. This is evaluated against the current state to determine if the transition is possible.

Reply:
Some context here - I initially split the `transition()` into two methods:
 - `prepareNextState(status: "intial" |  "input" | ... )` is responsible for creating the appropriate `State` object based on the status, it has the `switch` instruction inside
 - `transition(state: State)` takes in the `State` object, swaps the state and fires the `entry` and `exit` events
 
This is probably not the best way to do this though, I'm open for refactoring this as I'm doing bit of experimenting here